ISPN-7886 Cluster Counters over Hot Rod

pruivo opened the pull request:
https://issues.jboss.org/browse/ISPN-7886

Preview PR.

Missing parts:

- [x] listeners. I didn't check how to implement but probably I would like to have a single connection per client (?) instead of per counter (?). Does it make sense?
- [x] documentation. Message format in docs.
- [x] Add `CounterManager.getCounterNames()` support (hot rod operation) + documentation

ps. the operation codes skips `0x50` because it is assign to errors

NOTE: there are some random failure in the events that I need to debug :(

rvansa commented on server/hotrod/src/main/java/org/infinispan/server/hotrod/CacheDecodeContext.java:
```diff
@@ -75,6 +90,182 @@ public RequestParameters getParams() {
+
+   Response createCounter() {
+      CounterCreateDecodeContext decodeContext = operationContext();
+      boolean result = counterManager.defineCounter(decodeContext.getCounterName(), decodeContext.getConfiguration());
```
`defineCounter` involves RPC and is synchronous, while the decoder should not block. You need to add `defineCounterAsync` to the CounterManager.

pruivo replied:
I'm confused!
I don't think this method is invoked so often that it deserves an async approach. 
Also, all the cache operations are sync and use at least one RPC.

rvansa replied:
Ouch, the cache ops are really still sync! 😭 I thought this was already rewritten. It's not a matter of 'how often'. But due to other syncs, probably not really needed at this stage but the time shall come...

pruivo replied:
I created async method to the `EmbeddedCounterManager` but I kept the `CounterManager` API sync.  Only the `remove()` method is sync because I didn't find an easy way around the `ConcurrentHashMap.compute()`

rvansa commented on server/hotrod/src/test/java/org/infinispan/server/hotrod/counter/StrongCounterAPITest.java:
```diff
@@ -0,0 +1,111 @@
+ * @since 9.2
+ */
+@Test(groups = "functional", testName = "server.hotrod.counter.StrongCounterAPITest")
+public class StrongCounterAPITest extends HotRodMultiNodeTest {
```
I am not that familiar with client vs. server testsuite, but could you have a common base for tests in client & server? I am not fond of duplicities and TestCounters with non-trivial code just for the sake of tests either... @mgencur Do you know if this has been discussed before?

mgencur replied:
No. I don't think the relation between hotrod-client and hotrod-server test suite has been discussed. I'd agree that when the tests are very similar they should have a common base class in the hotrod-server module.

pruivo replied:
I don't think there are any common base test suite but you're right (at least for counters). The tests are the same and it is only testing the protocol (if everything is encoded/decoded correctly).

pruivo replied:
I managed to share the test between server and client. I don't think it would won a beauty contest but it seems to work :)

pruivo commented on client/hotrod-client/src/main/java/org/infinispan/client/hotrod/RemoteCacheManager.java:
```diff
@@ -352,6 +358,10 @@ public void close() throws IOException {
       stop();
    }
 
+   public CounterManager getCounterManager() {
```
TBH, I don't know. I miss the API meeting and the minutes talk about a RemoteInfinispan entry point. So I did the simplest thing :P
I'll change to `RemoteCounterManagerFactory` approach :)

Do you want to move the code out from infinispan-hotrod-client? 

I'll also shift the operations code to avoid conflict with the multimap!

rvansa replied:
I don't know if you should do that, especially since we don't have a concept of modularity in Hot Rod protocol, so it seems to make sense here. Also given then discussed discoverability of the API. Regrettably the only takeaway from the discussion seemed to me a vague "we need to play with the API and see..."

pruivo replied:
Added RemoteCounterManagerFactory to keep it symmetric with embedded counter and remote-multimap.
The code was kept in client-hotrod

pruivo commented on client/hotrod-client/src/main/java/org/infinispan/client/hotrod/counter/operation/CompareAndSetOperation.java:
```diff
@@ -0,0 +1,58 @@
+ * @author Pedro Ruivo
+ * @since 9.2
+ */
+public class CompareAndSetOperation extends BaseCounterOperation<Boolean> {
```
Sorry, I didn't understand what you mean :( there is a CAS operation in the protocol...

rvansa replied:
compare-and-*swap*

pruivo replied:
I'm not sure if you can convince me. TBH, I don't think the existing compare-and-set fits the counter logic and I'm sorry to have it added.

rvansa replied:
What problem do you have with CAS? (be it set or swap) Some application problems are already handled by the bounds, but I find the loop
```
do {
  long oldValue = counter.get();
  long newValue = businessLogic(oldValue);
} while (counter.compareAndSet(oldValue, newValue));
```
integral to any non-trivial use of shared counters. Compare-and-swap would add less RPCs:
```
long oldValue = counter.get();
long currentValue, newValue;
do {
   currentValue = oldValue;
   long newValue = businessLogic(oldValue);
} while ((oldValue = counter.compareAndSwap(oldValue, newValue)) != currentValue);
```

pruivo replied:
I don't know. It just looks like an AtomicLong use case than a Counter use case...

pruivo replied:
so... I guess I'll have to add the `compareAndSwap()` to the `StrongCounter` API in this PR...

rvansa replied:
I think about StrongCounter as distributed AtomicLong, don't you?

pruivo replied:
It has the same semantic but I don't see the strong-counter as an atomic-long. I guess we can advantage of it and improve the strong-counter API. 
I'm gonna create a JIRA for compare-and-swap.

pruivo commented on client/hotrod-client/src/main/java/org/infinispan/client/hotrod/counter/operation/CompareAndSetOperation.java:
```diff
@@ -0,0 +1,58 @@
+   private void assertBoundaries(short status) {
+      if (status == NOT_EXECUTED_WITH_PREVIOUS) {
+         if (update >= counterConfiguration.upperBound()) {
+            throw new CounterOutOfBoundsException(
```
it is correct. I use the `NOT_EXECUTED_WITH_PREVIOUS` to signal counter out of bound exception. I just identifying if it was the lower or upper bound to avoid return more data in the network.

rvansa replied:
My bad.

pruivo commented on client/hotrod-client/src/test/java/org/infinispan/client/hotrod/counter/CounterManagerTest.java:
```diff
@@ -0,0 +1,44 @@
+ * @since 9.2
+ */
+@Test(groups = "functional", testName = "client.hotrod.counter.CounterManagerTest")
+public class CounterManagerTest extends AbstractCounterTest {
```
I don't think it is needed but I can added it if it makes you happy. 
This kind of tests are already done in the counter module. Here I just test the protocol and see if the correct method is invoked.

pruivo commented on commons/src/main/java/org/infinispan/counter/api/CounterConfiguration.java:
```diff
@@ -1,16 +1,20 @@
+import static org.infinispan.counter.util.EncodeUtil.decodeStorage;
+import static org.infinispan.counter.util.EncodeUtil.decodeType;
+import static org.infinispan.counter.util.EncodeUtil.encodeTypeAndStorage;
+
```
the decode will fail with illegal state exception.

In the flags byte (probably needs a better name), the first 4 bits are the storage and the remaining 4 bits the type
`0000|0001` WEAK
`0000|0010` BOUNDED
`0000|0100` UNBOUNDED

`0001|0000` VOLATILE
`0010|0000` PERSISTENT

I hope this replies the question "where 0x80?" :)

rvansa replied:
Okay, I meant an earlier failure, but it's good enough.

Fine, I understand the `0x08`, but why do you use different bits (as mask) when those are just different values? Naturally

```
        v-- weak/strong
0000|0000
   ^   ^--- bounded/unbounded
   |------- volatile/persistent
```
which would give you 0 -> strong unbounded, 1 -> weak, 2 -> strong bounded, 3 -> illegal, ... 18 -> strong bounded persistent etc.

pruivo replied:
I changed the volatile/persistent to the 3rd bit. added a comment in EncodeUtils

pruivo commented on server/hotrod/src/main/java/org/infinispan/server/hotrod/CacheDecodeContext.java:
```diff
@@ -75,6 +90,182 @@ public RequestParameters getParams() {
+      }
+   }
+
+   void counterAddAndGet(Consumer<Response> sendResponse) {
```
what do you have in mind?

rvansa replied:
These all share the structure: fetch configuration, switch on type... `applyCounter(BiConsumer<StrongCounter, Consumer<Response>> applyStrong, BiConsumer<WeakCounter, Consumer<Response>> applyWeak)`

pruivo replied:
I was afraid someone would complain about too much functional methods :P
I applied your suggestion

pruivo commented:
Major update
* changed the encode
* added async method to EmbeddedCounterManager
* shift the operation code to avoid conflict with multimap (I think we are running out operation codes)
* refactor the test suite (it is shared between server and client)
* refactor CacheDecodeContext counter's methods

pruivo commented:
updated! added documentation!

tristantarrant commented:
Remote cache listeners currently have a severe limitation: they consume the connection on which they register (https://issues.jboss.org/browse/ISPN-8501).

pruivo commented:
@tristantarrant I'm going to try to avoid that. my plan is to use a single connection in the RemoteCounterManager and all the events are received in that connection.

pruivo commented:
Updated! I think I fixed the unordered events!

rvansa commented on client/hotrod-client/src/test/java/org/infinispan/client/hotrod/counter/BaseCounterAPITest.java:
```diff
@@ -0,0 +1,204 @@
+      T counter = counters.get(0);
+
+      Handle<BaseCounterImplTest.EventLogger> handle = addListenerTo(counter, new BaseCounterImplTest.EventLogger());
+      Thread.sleep(1000);
```
:-( ?

pruivo replied:
just to give some time to collect some events

rvansa commented on client/hotrod-client/src/test/java/org/infinispan/client/hotrod/counter/RemoteCounterManagerTest.java:
```diff
@@ -0,0 +1,102 @@
+
+
+   public RemoteCounterManagerTest() {
+      testImpl = new CounterManagerImplTest(this::allTestCounterManagers, this::log, this::cacheManager);
```
fyi, you can use `@Factory` method to replace the actually tested instance (its methods have to be annotated with `@Test` ofc)

pruivo replied:
IMO, this is simpler instead of trying to adapt it to use the `@Factory`

danberindei replied:
@pruivo I don't like the idea of having classes that end in `Test` but aren't really tests. A test strategy maybe?

+1 to use `@Factory` or even inheritance, we shouldn't bend ourselves backwards just because TestNG is weird about how groups are inherited.

pruivo replied:
If I use the `@Factory` annotation, it fails
```
java.lang.IllegalStateException: Test org.infinispan.client.hotrod.counter.WeakCounterAPITest extends MultipleCacheManagersTest and declares its own @Factory method: 
```
If I extend the `factory()` method as 
```
   @Override
   public Object[] factory() {
      return new Object[]{
            new WeakCounterImplTest(),
            new WeakCounterAPITest()
      };
   }
```
the tests aren't invoked and they don't run.

pruivo replied:
@danberindei AFAIK, I can't extend 2 abstract classes :)

danberindei replied:
@pruivo That's easy to fix, don't extend `AbstractCounterTest` :)
Although I never liked extending tests from one module in 

Re: `@Factory`, yes, you're not supposed to use it in subclasses of `MultipleCacheManagersTest`. We could probably allow it, only I'm not 100% sure how TestNG handles multiple `@Factory` annotations so I'd rather make all the tests use `@Factory` and skip the tests explicitly instead of using the `@InCacheMode` and similar annotations

Also, I recently (ISPN-8537) changed `factory()` to check for copy-paste errors and require the returned objects to be instances of the concrete test class. Unfortunately I lost the `@Test` annotation in one of my iterations and now "wrong" instance types are ignored instead of failing explicitly. I will have a PR soon.

pruivo replied:
-1 I don't want to copy the `MultiHotRodServersTest` 3 times... (I'm running 3 tests from the server module)
I'll change the name to test strategy as suggested above

rvansa commented on client/hotrod-client/src/test/java/org/infinispan/client/hotrod/counter/StrongCounterAPITest.java:
```diff
@@ -0,0 +1,117 @@
+
+   @Override
+   public void testExceptionInListener(Method method) throws InterruptedException {
+      super.testExceptionInListener(method);
```
why do you need to override this?

pruivo replied:
the `StrongCounterTest` is an interface and this way I make sure that I'll run all the tests :)

rvansa replied:
I thought that this is what the interface enforces...

pruivo replied:
nevermind, I was looking at another test :) 
I didn't mark the base class as test and I need to override here.

danberindei replied:
IIRC you could put the `@Test` annotation on the super method.

rvansa commented on commons/src/main/java/org/infinispan/counter/util/EncodeUtil.java:
```diff
@@ -0,0 +1,160 @@
+    * <p>
+    * See the documentation for further details about the encoding.
+    * <p>
+    * TODO! link to docs
```
todo

pruivo replied:
there is no docs yet :)

pruivo replied:
I've added a link to where the docs should be :)

rvansa commented on counter/src/main/java/org/infinispan/counter/impl/CounterModuleLifecycle.java:
```diff
@@ -165,10 +168,18 @@ public void cacheManagerStarted(GlobalComponentRegistry gcr) {
+         //local only cache manager. unable to create replicated or distributed caches!
+         future = new CompletableFuture<>();
+         future.completeExceptionally(log.expectedClusteredEnvironment());
+         //future.completeExceptionally(new CounterException("Clustered counters only available when clustered cache manager."));
```
remove commented code

rvansa commented on counter/src/main/java/org/infinispan/counter/impl/manager/CacheHolder.java:
```diff
@@ -56,15 +59,16 @@ void registerNotificationManager(CounterManagerNotificationManager notificationM
+                  return CompletableFuture.completedFuture(fileConfig);
+               }
+               //put configuration in cache
+               return configCacheWithFlags(configuration).putIfAbsentAsync(name, configuration);
```
I don't follow; aren't you calling the `putIfAbsentAsync` in `checkAndStoreConfiguredCounterConfigurationAsync` already?

pruivo replied:
`checkAndStoreConfiguredCounterConfigurationAsync` only checks the counters defined in the configuration file. if the counter exists, I put it in the cache (`putIfAbsent` in the method)
if it is not defined in the configuration, we put the in the cache with the user's `CounterConfiguration`.

pruivo replied:
no it is lazy since day 2 :) the configuration is added to the cache when they are accessed for the first time.

pruivo commented on client/hotrod-client/src/main/java/org/infinispan/client/hotrod/counter/impl/CounterOperationFactory.java:
```diff
@@ -0,0 +1,98 @@
+      this.configuration = configuration;
+      this.transportFactory = transportFactory;
+      this.codec = codec;
+      this.topologyId = transportFactory.createTopologyId(RemoteCacheManager.cacheNameBytes("org.infinispan.counter"));
```
what other instance?

pruivo replied:
the cache is private and it shouldn't be acessible via Hot Rod. And I don't wanna expose the 
internals of counters so they can be changed in the future without have to worries with compatiblity.

rvansa replied:
OK, let's address it separately.

pruivo commented on client/hotrod-client/src/main/java/org/infinispan/client/hotrod/counter/impl/CounterHelper.java:
```diff
@@ -0,0 +1,40 @@
+ * @author Pedro Ruivo
+ * @since 9.2
+ */
+class CounterHelper {
```
this is a no go. java doesn't "detect" the default methods in `CounterTrait` and fails to compile. I have to delegate everything anyway.

rvansa replied:
OK, in order to work the `CounterTrait` would has to implement the interfaces as well, which is not possible in this case.

pruivo commented on client/hotrod-client/src/main/java/org/infinispan/client/hotrod/counter/operation/BaseCounterOperation.java:
```diff
@@ -0,0 +1,73 @@
+
+   BaseCounterOperation(Codec codec, TransportFactory transportFactory, AtomicInteger topologyId, Configuration cfg,
+         String counterName) {
+      super(codec, transportFactory, EMPTY_CACHE_NAME, topologyId, 0, cfg);
```
The counters operation can be invoked in any node in the cluster. However, the TransportFactory needs a cache name. I didn't want to add complexity and implement a counter transport factory so I'm setting the cache name to update the list of servers in the factory.

pruivo replied:
1st, `TransportFactory.getTransport(Set<SocketAddress> failedServers, byte[] cacheName);` < needs the cache name!
2nd, I don't want to expose any internals. If we change the counters to use anything else than a cache, I don't want to break the clients! 
3rd, if an use implements the protocol in its own language, I don't want them to know that they suppose to use a specific cache name.

rvansa replied:
My bad, I haven't noticed that you don't use the cache name in the request header, only when reading the response. Besides not exposing the internals, it makes sense to save those bytes in the request.
It would make sense to drop the cache name from the protocol completely, though the complexity of parsing (based on opCode as flags come after cache name) is probably not worth the saved single byte.

pruivo replied:
I prefer to have the cache name there too.

pruivo commented on client/hotrod-client/src/main/java/org/infinispan/client/hotrod/impl/protocol/Codec20.java:
```diff
@@ -466,4 +511,15 @@ protected void readNewTopologyAndHash(Transport transport, HeaderParams params)
+      short status = transport.readByte();
+      assert status == 0;
+      short topology = transport.readByte();
+      assert topology == 0;
```
I'm using the event header as http://infinispan.org/docs/dev/user_guide/user_guide.html#remote_events
so the status and topology is already 0

tristantarrant commented:
Pedro, lots of failures caused by:
`ERROR [org.jboss.msc.service.fail] (MSC service thread 1-1) MSC000001: Failed to start service jboss.datagrid-infinispan-endpoint.hotrod.hotrod-connector: org.jboss.msc.service.StartException in service jboss.datagrid-infinispan-endpoint.hotrod.hotrod-connector: Failed to start service
	at org.jboss.msc.service.ServiceControllerImpl$StartTask.run(ServiceControllerImpl.java:1904)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
Caused by: java.lang.NoClassDefFoundError: org/infinispan/counter/EmbeddedCounterManagerFactory
	at org.infinispan.server.hotrod.HotRodServer.startInternal(HotRodServer.java:213)
	at org.infinispan.server.hotrod.HotRodServer.startInternal(HotRodServer.java:94)
	at org.infinispan.server.core.AbstractProtocolServer.start(AbstractProtocolServer.java:78)
	at org.infinispan.server.endpoint.subsystem.SecurityActions$6.run(SecurityActions.java:136)
	at org.infinispan.server.endpoint.subsystem.SecurityActions$6.run(SecurityActions.java:133)
	at org.infinispan.security.Security.doPrivileged(Security.java:76)
	at org.infinispan.server.endpoint.subsystem.SecurityActions.doPrivileged(SecurityActions.java:42)
	at org.infinispan.server.endpoint.subsystem.SecurityActions.startProtocolServer(SecurityActions.java:140)
	at org.infinispan.server.endpoint.subsystem.ProtocolServerService.startProtocolServer(ProtocolServerService.java:194)
	at org.infinispan.server.endpoint.subsystem.ProtocolServerService.start(ProtocolServerService.java:152)
	at org.jboss.msc.service.ServiceControllerImpl$StartTask.startService(ServiceControllerImpl.java:1948)
	at org.jboss.msc.service.ServiceControllerImpl$StartTask.run(ServiceControllerImpl.java:1881)
	... 3 more
Caused by: java.lang.ClassNotFoundException: org.infinispan.counter.EmbeddedCounterManagerFactory from [Module "org.infinispan.server.hotrod:main" from local module loader @1c2c22f3 (finder: local module finder @18e8568 (roots: /home/infinispan/workspace/Infinispan_PR-5552-SDH5OGXHT5REXE4KH43R2AH2U76XONZHQ7QYE3HWU52JSS5C2JAA/server/integration/testsuite/target/server/node2/modules,/home/infinispan/workspace/Infinispan_PR-5552-SDH5OGXHT5REXE4KH43R2AH2U76XONZHQ7QYE3HWU52JSS5C2JAA/server/integration/testsuite/target/server/node2/modules/system/layers/base))]
	at org.jboss.modules.ModuleClassLoader.findClass(ModuleClassLoader.java:198)
	at org.jboss.modules.ConcurrentClassLoader.performLoadClassUnchecked(ConcurrentClassLoader.java:363)
	at org.jboss.modules.ConcurrentClassLoader.performLoadClass(ConcurrentClassLoader.java:351)
	at org.jboss.modules.ConcurrentClassLoader.loadClass(ConcurrentClassLoader.java:93)
	... 15 more`

pruivo commented:
@tristantarrant should be fixed now! thanks!

ryanemerson commented:
Integrated, thanks @pruivo

danberindei commented on server/hotrod/src/main/java/org/infinispan/server/hotrod/CacheDecodeContext.java:
```diff
@@ -75,6 +100,145 @@ public RequestParameters getParams() {
+   }
+
+   void getCounterConfiguration(Consumer<Response> sendResponse) {
+      ((EmbeddedCounterManager) counterManager).getConfigurationAsync(operationContext())
```
Might as well make the field an `EmbeddedCounterManager`.

pruivo commented on client/hotrod-client/src/main/java/org/infinispan/client/hotrod/RemoteCounterManagerFactory.java:
```diff
@@ -0,0 +1,20 @@
+   private RemoteCounterManagerFactory() {
+   }
+
+   public static CounterManager asCounterManager(RemoteCacheManager cacheManager) {
```
just to be consistent with `EmbeddedCounterManagerFactory.asCounterManager()`. In any case, I think the factories would be removed in the future and we are going to access counters/locks/etc. in other way.

pruivo commented on client/hotrod-client/src/main/java/org/infinispan/client/hotrod/counter/impl/ConnectionManager.java:
```diff
@@ -0,0 +1,217 @@
+   }
+
+   public synchronized void stop() {
+      removeConnection("");
```
an empty counter name in the server means remove all listeners associated to the id and drop the connection.

pruivo commented on client/hotrod-client/src/main/java/org/infinispan/client/hotrod/counter/impl/ConnectionManager.java:
```diff
@@ -0,0 +1,217 @@
+   synchronized void removeConnection(String counterName) {
+      if (trace) {
+         log.tracef("Remove listener connection for counter '%s'", counterName);
+      }
```
+.5 to braces. IMO, it makes the code more readable instead of being all bumped together.

pruivo commented on client/hotrod-client/src/main/java/org/infinispan/client/hotrod/counter/impl/ConnectionManager.java:
```diff
@@ -0,0 +1,217 @@
+   }
+
+   private String getThreadName() {
+      return "Counter-Listener-" + Util.toHexString(listenerId, 8);
```
I don't think so...

pruivo commented on client/hotrod-client/src/main/java/org/infinispan/client/hotrod/counter/impl/ConnectionManager.java:
```diff
@@ -0,0 +1,217 @@
+         registeredCounters.forEach(this::createConnection);
+      } catch (TransportException e) {
+         if (debug) {
+            log.debug("Unable to find another server, so ignore connection reset");
```
it retries later
```
factory.geTransportFactory().addDisconnectedListener(this::findAnotherServer);
```

pruivo commented on client/hotrod-client/src/main/java/org/infinispan/client/hotrod/event/ClientListenerNotifier.java:
```diff
@@ -328,7 +328,12 @@ private void tryFailoverClientListener() {
          } catch (TransportException e) {
             log.debug("Unable to failover client listener, so ignore connection reset");
```
yes... since my knowledge of hot rod and listener was zero, I tried to copy the most of it from the clustered listener :)

pruivo commented on counter/src/main/java/org/infinispan/counter/impl/listener/CounterManagerNotificationManager.java:
```diff
@@ -215,7 +215,11 @@ public void updateState(CacheEntryEvent<? extends CounterKey, CounterValue> even
+         synchronized (holder.generator) {
+            //weak counter events execute the updateState method in parallel.
+            //if we don't synchronize, we can have events reordered.
+            triggerUserListener(holder.userListeners, holder.generator.generate(key, event.getValue()));
```
this is safe. the user listener is invoked in a dedicated thread and not this thread.

pruivo commented on counter/src/main/java/org/infinispan/counter/impl/weak/WeakCounterImpl.java:
```diff
@@ -232,46 +238,71 @@ public String toString() {
+               value = Math.addExact(value, toAdd);
+            }
+         } catch (ArithmeticException e) {
+            return getCachedValue0(index, value, -1);
```
where did you copy that code? :)

pruivo commented on documentation/src/main/asciidoc/user_guide/server_protocols_hotrod_wire.adoc:
```diff
@@ -2180,6 +2182,467 @@ The XID in the requests has the following format:
+* `0x00`: Operation successful. +
+* `0x01`: Operation failed. +
+* `0x02`: The counter isn't defined. +
+* `0x04`: The counter reached a boundary. Only possible for `STRONG` counters.
```
there is no `BOUND` counters. they are weak or strong. and the strong can have boundaries.
